lib/elixir/lib/string_io.ex

```diff
@@ -359,6 +337,25 @@ defmodule StringIO do
 
   ## helpers
 
+  defp capture_prompt(%{capture_prompt: false} = s, _prompt), do: s
+  defp capture_prompt(%{capture_prompt: true, output: output} = s, prompt) do
+    %{s | output: <<output::binary, IO.chardata_to_string(prompt)::binary>>}
+  end
+
+  defp binary_collect("", _, acc), do: {acc, ""}
+  defp binary_collect(<<"\r\n"::binary, tail::binary>>, _, acc), do: {<<acc::binary, "\n">>, tail}
+  defp binary_collect(<<"\n"::binary, tail::binary>>, _, acc), do: {<<acc::binary, "\n">>, tail}
+
+  defp binary_collect(<<head::utf8, tail::binary>>, :unicode, acc) do
+    binary_collect(tail, :unicode, <<acc::binary, head::utf8 >>)
+  end
+
+  defp binary_collect(<<head, tail::binary>>, :latin1, acc) do
+    binary_collect(tail, :latin1, <<acc::binary, head >>)
+  end
+
+  defp binary_collect(<<_::binary>>, _, _), do: :error
```

Comment:
I'm pretty sure it should be enough to just count the characters and do a `:erlang.split_binary` at the end. This way we can avoid copying.